lib/Cro/ResourceIdentifier.pm6

```diff
@@ -0,0 +1,57 @@
+#| The kind of host found in the URI (a domain name or some kind of IP address)
+enum Cro::ResourceIdentifier::Host <RegName IPv4 IPv6 IPvFuture>;
+
+role Cro::ResourceIdentifier {
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return "https"
+    has Str $.scheme;
+
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return "user@cro.services:44433"
+    has Str $.authority;
+
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return "user"
+    has Str $.userinfo;
+
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return "cro.services"
+    has Str $.host;
+
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return Cro::Uri::Host::RegName
+    has Cro::ResourceIdentifier::Host $.host-class;
+
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return 44433
+    has $.port;
+
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return "/example/url"
+    has Str $.path;
+
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return "foo=bar&x=42"
+    has Str $.query;
+
+    #| Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return "here"
+    has Str $.fragment;
+
+    #| Return an array of path segments, decoding any percent sequences found in
+    #| them. Given the example "https://user@cro.services:44433/example/url?foo=bar&x=42#here",
+    #| this would return an array ["example", "foo"].
+    method path-segments(Cro::ResourceIdentifier:D: --> List) {
+        my $no-leader = $!path.starts-with('/') ?? $!path.substr(1) !! $!path;
+        $no-leader.split('/').map(&decode-percents).list
+    }
+}
+
+sub decode-percents(Str $s) {
```

Comment:
That's obviously a dupe that shouldn't be here.

Reply:
I think it's too big a back-compat break, *but* we could have them in there, and then re-export them from `Cro::Uri` (the `package EXPORT::tag-name { our &decode-percents := ...; }` trick).

Reply:
Ah, and "What about the except-ASCII variant?" - I don't know, maybe it is useful on its own also..